dict.go

```diff
@@ -56,7 +62,9 @@ func (d *Dictionary) postingsList(term []byte, except *roaring.Bitmap, rv *Posti
 		return d.postingsListInit(rv, except), nil
 	}
 
+	d.m.Lock()
```

Comment:
can you please explain why this lock is necessary? 
from what i know this path is hit at query time after all the fsts and their readers have been loaded, so at this point all the threads that do access this segment and try to form a postings list would be reading the dictionary - so i dont think a lock is necessary. and there wouldn't be a write - read conflict because the reader is populated after loading the dictionary from the sb.mem[], so this reader's read API won't conflict with the creating and populating the reader inside the dictionary. please let me know if this makes sense

Reply:
Yes that is true
the fstReader is meant for single threaded use, and hence so is the dictionary object
To accomodate this in the query path we essentially create a dictionary object for the field queried from all segments - So for N segments, N dictionary objects and N fst Iterators, because each dictionary create creates a fresh FST Iterator Object.

But the new use case i added is that you cache the dictionary for the "_id" field in the segment - thereby making it a shared object that can be subject to multithreaded access, in which case the lock is needed to allow for single threaded access to the cached "_id" dict's fst reader

This is beacuse from the pprof i could see that for every segment introduced, in a high indexing op scenario,

for an existing segment object S, S's DocNumbers function was called in the introducer, on an avg of around 20 times (after that the segment S just gets deleted due to merge). So the same id dict gets created 20 times, with 20 different FST iterator objects being created as well, which creates a lot of garbage to be GC'd.

Reply:
hmm ok, thing is in the query path the dictionary object is never shared across multiple threads, so the lock is useless in that path.
So i can actually create a new function called postingsListLock() that will only be called by DocNumbers() in the introducer path which is the only place where a doctionary can be shared across threads?

Reply:
I see yeah now I see that only the idDict that's cached would be shared across threads. 

I think even DocID searcher uses docNumbers() API and if there are parallel deployment of those queries - and while they form the https://github.com/blevesearch/bleve/blob/5f1f45a5c32a140546e247ba91e7250de247adf9/index/scorch/snapshot_index.go#L331 

object parallely there can be shared variable contention over here as well? Won't affect the other types of queries that are there - but in this particular query there can be contention?

Reply:
yes multiple docID queries would use the same dict, so the lock would come into effect there, so then i could seperate this as well, docID queries will call the old API and introducer will call DocNumbersShared, only difference being that from query side (DocIDquery) a new dict is always created and not cached